ws/WebSocketClient.ts

```diff
@@ -1,127 +1,223 @@
+  disconnect(code?: number) {
+    if (code === 1000 || code === 1001) {
+      this.sessionID = null;
+      this._seq = null;
     }
+    this._selfDisconnect = true;
+    this.ws?.close(code);
+    return this;
   }
 
-  private _heartbeat(expected: boolean) {
-    if (expected) {
-      if (!this._lastHeartbeatAck) {
-        this._terminate();
-        this.connect();
-        this._resume();
-        this._lastHeartbeatAck = true;
-      } else {
-        this._lastHeartbeatAck = false;
-      }
+  // TODO prevent initialization without url
+  initialize() {
+    this._selfDisconnect = false;
+    if (this.ws) throw new Error('Attempted duplicate connection - run uninitialize');
+    this.ws = new WebSocket(this._url);
+    this.ws.once('close', this._onClose);
+    this.ws.on('message', this._onMessage);
+    this.ws.on('error', console.error);
+    return this;
+  }
+
+  restart(newSession = false) {
+    if (newSession) this.sessionID = null;
+    this.disconnect().connect();
+  }
+
+  // TODO Prevent 4012, 4013?
+  send(
+    op: GATEWAY_OPCODES.IDENTIFY | GATEWAY_OPCODES.RESUME | GATEWAY_OPCODES.HEARTBEAT | GATEWAY_OPCODES.REQUEST_GUILD_MEMBERS | GATEWAY_OPCODES.VOICE_STATE_UPDATE | GATEWAY_OPCODES.PRESENCE_UPDATE,
+    d: any,
+    override = false,
+  ) {
+    if (this.ws?.readyState !== WebSocket.OPEN) return false;
+    if (override
+      || (this.ws?.readyState === WebSocket.OPEN
+        && ((this.ready && !(op === GATEWAY_OPCODES.IDENTIFY || op === GATEWAY_OPCODES.RESUME))
+          || (!this.ready && (op === GATEWAY_OPCODES.IDENTIFY || op === GATEWAY_OPCODES.RESUME))
+        )
+      )
+    ) {
+      this.ws.send(JSON.stringify({ op, d }));
+      return true;
     }
-    this.send(GATEWAY_OPCODES.HEARTBEAT, this._seq || null);
+    return false;
   }
 
-  private _identify() {
-    this.send(GATEWAY_OPCODES.IDENTIFY, {
-      token: this._token,
-      intents: this._intents,
-      properties: {
-        $os: process.platform,
-        $browser: 'Dis.ts',
-        $device: 'Dis.ts',
-      },
-    });
+  setIntents(intents: number) {
+    this._intents = intents;
   }
 
-  private _onClose(code: number) {
-    if (!GATEWAY_CLOSE_EVENT_CODES[code]) console.log(code);
-    else console.log(GATEWAY_CLOSE_EVENT_CODES[code]);
+  setToken(token: string) {
+    this._token = token;
   }
 
-  private _onMessage(data: Payload) {
-    data = JSON.parse(String(data));
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    const { t, s, op, d } = data;
-    if (s) this._seq = s;
+  setURL(url: string) {
+    this._url = url;
+  }
 
-    switch (op) {
-      case GATEWAY_OPCODES.DISPATCH:
-        // if (EVENTS[t]) client.emit(EVENTS[t], d);
-        this._sessionID = d.session_id;
-        break;
+  uninitialize() {
+    if (this._heartBeatInterval) clearInterval(this._heartBeatInterval);
+    if (this.ws?.readyState === WebSocket.OPEN) this.ws.off('message', this._onMessage).off('error', console.error).terminate();
```

Comment:
maybe a little more verbose but I would at least multiline this and actually not chain it.